api/v1alpha5/conversion.go

```diff
@@ -16,26 +16,106 @@ limitations under the License.
 
 package v1alpha5
 
-// Hub marks OpenStackCluster as a conversion hub.
-func (*OpenStackCluster) Hub() {}
+import (
+	ctrlconversion "sigs.k8s.io/controller-runtime/pkg/conversion"
 
-// Hub marks OpenStackClusterList as a conversion hub.
-func (*OpenStackClusterList) Hub() {}
+	infrav1 "sigs.k8s.io/cluster-api-provider-openstack/api/v1alpha6"
+)
 
-// Hub marks OpenStackClusterTemplate as a conversion hub.
-func (*OpenStackClusterTemplate) Hub() {}
+var _ ctrlconversion.Convertible = &OpenStackCluster{}
 
-// Hub marks OpenStackClusterTemplateList as a conversion hub.
-func (*OpenStackClusterTemplateList) Hub() {}
+func (r *OpenStackCluster) ConvertTo(dstRaw ctrlconversion.Hub) error {
+	dst := dstRaw.(*infrav1.OpenStackCluster)
 
-// Hub marks OpenStackMachine as a conversion hub.
-func (*OpenStackMachine) Hub() {}
+	return Convert_v1alpha5_OpenStackCluster_To_v1alpha6_OpenStackCluster(r, dst, nil)
```

Comment:
For not loosing information during conversion, we have to unmarshal the data to get it added to the `cluster.x-k8s.io/conversion-data` Annotation (see current implementation in `v1alpha4` - https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/e78f0b83b0297590043826138b33f7c050f1ba53/api/v1alpha4/conversion.go#L32-L44)

This must be also done in the `ConvertFrom` method below (see current implementation in `v1alpha4` - https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/e78f0b83b0297590043826138b33f7c050f1ba53/api/v1alpha4/conversion.go#L52-L59)

This procedure applies also to `ConvertTo`/`ConvertFrom` methods for `OpenStackMachine` and `OpenStackMachineTemplate`.

Unfourtanetly i just discovered, that i didn't implemented the `utilconversion.UnmarshalData` in `ConvertTo` for `OpenStackMachine` and `OpenStackMachineTemplate` when i introduced this in https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1247.

Reply:
That's an interesting and nasty thing, tbh. Are we able to unit test this somehow? What data is going to be lost without the (un)marshalling?

Reply:
That is really a nasty thing, you're toally right. Thanks! Let me implement this for these 3 types.

Reply:
Are we able to test this somehow?

Reply:
> Are we able to test this somehow?

if you already have a cluster where `v1alpha6` is applied, you could do a `k get openstackcluster.v1alpha5.infrastructure.cluster.x-k8s.io -o yaml`. This should give you beside the spec of the object in version `v1alpha5` in the `cluster.x-k8s.io/conversion-data` annotation the same spec in version `v1alpha6`.

Reply:
I also added the marshalling fix for `OpenStackClusterTemplate`